fendermint/vm/topdown/src/sync.rs

```diff
@@ -0,0 +1,147 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+//! A constant running process that fetch or listener to parent state
+
+use crate::error::Error;
+use crate::{BlockHeight, Config, ParentFinalityProvider};
+use anyhow::{anyhow, Context};
+use ipc_sdk::cross::CrossMsg;
+use ipc_sdk::ValidatorSet;
+use std::cmp::max;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use std::time::Duration;
+
+/// Constantly syncing with parent through polling
+pub struct PollingParentSyncer<T> {
+    config: Config,
+    started: Arc<AtomicBool>,
+    parent_view_provider: Arc<T>,
+    agent: Arc<IPCAgentProxy>,
+}
+
+impl<T> PollingParentSyncer<T> {
+    pub fn new(config: Config, parent_view_provider: Arc<T>, agent: Arc<IPCAgentProxy>) -> Self {
+        Self {
+            config,
+            started: Arc::new(AtomicBool::new(false)),
+            parent_view_provider,
+            agent,
+        }
+    }
+}
+
+impl<T: ParentFinalityProvider + Send + Sync + 'static> PollingParentSyncer<T> {
+    /// Start the proof of finality listener in the background
+    pub fn start(&mut self) -> anyhow::Result<()> {
+        match self
+            .started
+            .compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst)
+        {
+            Ok(_) => {}
+            Err(_) => return Err(anyhow!("already started")),
+        }
```

Comment:
```suggestion
    pub fn start(self) {    
```

By consuming `self` this check can be completely avoided. You set up the component, then run it till the end of the program, optionally returning some handle to stop it. We communicate with the rest of the world through the parent view anyway, so there doesn't seem to be a need to call this once it's started.

Reply:
^+1